### ISSUE TITLE
Change (primarily Operator) nodes to have more descriptive names

### ISSUE THREAD

**Issue by Zyzyl** [Frooxius/NeosPublic#795]
I and others have variously mentioned in Discord that the current names for several nodes could benefit from being more descriptive. I realize many of the current names are standard symbols in many programming languages and thus may appear very familiar to those with software engineering experience. However Neos is not aimed solely at a software engineer audience and renaming some nodes would help lower the barrier to entry for using one of the main Neos creation systems.

Specifically I suggest making changes along these lines:
- ! becomes "Is Not !" or "Not !"
- != becomes "Is Not Equal To !=" or "Not Equal !="
- % becomes "Remainder %" or "Modulo %"
- & becomes "And &"
- ?: becomes "Ifelse ?:"
- ?? becomes... something better, I'm not 100% sure what this actually does
- | becomes "Or |"

Also:
- -> becomes "Dereference ->" or something else if I've got the wrong technical term here.

There are several others which could probably benefit from the same treatment, but these seem like the most important. Particularly !, !=, &, ?:, and | since these are all nodes where the use case is pretty clear if the node is named in plain language but potentially completely opaque if not. :? is probably the worst case here since it's such a useful node but I can't overstate how unhelpful the current name is for most people.

A secondary point which affects some of the mentioned nodes + a few others is that wiki pages absolutely cannot contain the characters `# < > [ ] { } |` (see [here](https://en.wikipedia.org/wiki/Wikipedia:Naming_conventions_(technical_restrictions)#Forbidden_characters)). Adding some descriptive text to the name of any nodes containing these would make it easier for people to search for the node's wiki page.

Finally, I suggest making bools the default input type for !, & and |. I can add that as a comment here if you prefer https://github.com/Frooxius/NeosPublic/issues/358#issue-546735215

**Comment by Frooxius** [comment]
I think we can do both. My long term goal was to actually have a simple graphic on each node that would have the main symbol and then smaller text for the word description. Maybe even some simple infographic.

The problem with longer names (and why I used symbols in the first place) is that they're a bit harder to read and slower to parse. Especially if you're using a VR headset (particularly older one with lower resolution), you'd need to get close to the nodes to see what they are, whereas symbols are recognizable from afar.

Having a big symbol and smaller text (e.g. ! with smaller "Not" underneath) could work well. If you're not sure what it is, you look up close and you remember the symbol at the same time, making the work easier later on.

I'm not sure about ?: being "If Else" though, as that would conflict a bit with the flow nodes which already has "If Else" (although it only says "If").

?? is also problematic. It's called null coalescing operator and I don't know if there's a short description for it that would fit on the node. This is a point where having a direct link to documentation for the nodes would help instead.

**Comment by mralext20** [comment]
The conflict of if and :? with flow nodes seemed like a non issue, considering the existing of ++ and -- both in the actions folder and in operators

**Comment by mralext20** [comment]
?? To `if not null else`

**Comment by Zyzyl** [comment]
Thanks Froox, I think adding some smaller text for these while retaining the large symbol would be a reasonable approach.

Perhaps ?: could be called "Conditional" or similar, though I don't think "Ifelse would be too much of an issue (or unprecendented as Alex pointed out). Just, looked ?? up - maybe that could be "First Not Null" or similar?

As and when you decide to implement this I would suggest making the changes to pretty much all nodes with only non-alphabetical names, _maybe_ + - x ÷ could be spared.

**Comment by Frooxius** [comment]
@mralext20 It's not as much about the name clash, as that the ++ and -- do pretty much the same thing, but ?: is a bit different semantically, as it deals with evaluation, rather than program flow.

Conditional sounds good though, that's short and simple enough. It's often named conditional or ternary operator so that would fit.

"First not null" might work yeah. The smaller labels could be added to all anyways, the ÷ confuses a few people too, so if it has small "Div" or "Divide" under that can help.

**Comment by Earthmark** [comment]
I do agree that more verbose names would be nice, there has been many times where I'm explaining logix to someone and they're like "dafuq's this ?: thing?", And when I explain it and they ask why all I can really say is "it's a c/c++/c# thing..." And I realize I don't really have a good answer.
The same thing has happened with or, and, not, and xor. Oddly the Boolean ops that are spelled out still confuse people, but they don't seem to care as much?

**Comment by sirkitree** [comment]
I'd request that we do not make these logix nodes larger with longer text, but have a way to have an explanation on hover or something. LogiX already can get large and hard to parse over an area, and the longer each node is the harder it is to make it compact and parse quickly in my head. But having longer descriptions that are only shown on hover could be pretty decent.

**Comment by Zyzyl** [comment]
@sirkitree That also sounds like a possible option, although I think the additional explanatory text would absolutely need to be visible in the node browser without spawning the node.

I don't think node sizes would necessarily grow muich though - there is often unused space on the node visual and Froox was suggesting using a smaller font for the more descriptive name anyway.

**Comment by Frooxius** [comment]
@Earthmark It's not symbols specifically because of C/C++/C# thing, but to help improve readability. Consider headsets that have lower resolutions, for longer names you have to get much closer to them to read what the node is.

However the current symbols are tentative solution as well, the ultimate goal is to have a proper graphic for each of those core nodes, that combines both approaches.

@sirkitree That could work too, although I think a small description would fit in the node as well, while keeping the existing symbol at pretty much the same size. We'll have to experiment though.

I definitely want to maintain the ability to quickly parse and read what the nodes are though. 

And yeah the description would need to be show in node browser too, but that shouldn't be a problem.

**Comment by Banane9** [comment]
> 
> 
> I'd request that we do not make these logix nodes larger with longer text, but have a way to have an explanation on hover or something. LogiX already can get large and hard to parse over an area, and the longer each node is the harder it is to make it compact and parse quickly in my head. But having longer descriptions that are only shown on hover could be pretty decent.

Very much agree with @sirkitree there - the symbol nodes are recognizable as a programmer, and quick to read.  
Descriptions on hover are lacking all over the game, really.
